File: examples/matlab/PlanarSLAMExample_easy.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% GTSAM Copyright 2010, Georgia Tech Research Corporation, 
% Atlanta, Georgia 30332-0415
% All Rights Reserved
% Authors: Frank Dellaert, et al. (see THANKS for the full author list)
% 
% See LICENSE for the license information
%
% @brief Simple robotics example using the pre-built planar SLAM domain
% @author Alex Cunningham
% @author Frank Dellaert
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Assumptions
%  - All values are axis aligned
%  - Robot poses are facing along the X axis (horizontal, to the right in images)
%  - We have bearing and range information for measurements
%  - We have full odometry for measurements
%  - The robot and landmarks are on a grid, moving 2 meters each step
%  - Landmarks are 2 meters away from the robot trajectory

%% Create keys for variables
x1 = 1; x2 = 2; x3 = 3;
l1 = 1; l2 = 2;

%% Create graph container and add factors to it
graph = PlanarSLAMGraph;

%% Add prior
% gaussian for prior
prior_model = SharedDiagonal([0.3; 0.3; 0.1]);
prior_measurement = Pose2(0.0, 0.0, 0.0); % prior at origin
graph.addPrior(x1, prior_measurement, prior_model); % add directly to graph

%% Add odometry
% general noisemodel for odometry
odom_model = SharedDiagonal([0.2; 0.2; 0.1]);
odom_measurement = Pose2(2.0, 0.0, 0.0); % create a measurement for both factors (the same in this case)
graph.addOdometry(x1, x2, odom_measurement, odom_model);
graph.addOdometry(x2, x3, odom_measurement, odom_model);

%% Add measurements
% general noisemodel for measurements
meas_model = SharedDiagonal([0.1; 0.2]);

% create the measurement values - indices are (pose id, landmark id)
degrees = pi/180;
bearing11 = Rot2(45*degrees);
bearing21 = Rot2(90*degrees);
bearing32 = Rot2(90*degrees);
range11 = sqrt(4+4);
range21 = 2.0;
range32 = 2.0;

% create bearing/range factors and add them
graph.addBearingRange(x1, l1, bearing11, range11, meas_model);
graph.addBearingRange(x2, l1, bearing21, range21, meas_model);
graph.addBearingRange(x3, l2, bearing32, range32, meas_model);

% print
graph.print('full graph');

%% Initialize to noisy points
initialEstimate = PlanarSLAMValues;
initialEstimate.insertPose(x1, Pose2(0.5, 0.0, 0.2));
initialEstimate.insertPose(x2, Pose2(2.3, 0.1,-0.2));
initialEstimate.insertPose(x3, Pose2(4.1, 0.1, 0.1));
initialEstimate.insertPoint(l1, Landmark2(1.8, 2.1));
initialEstimate.insertPoint(l2, Landmark2(4.1, 1.8));

initialEstimate.print('initial estimate');

%% Optimize using Levenberg-Marquardt optimization with an ordering from colamd
result = graph.optimize_(initialEstimate);
result.print('final result');

%% Print out the corresponding dense matrix
ord = graph.orderingCOLAMD(result);
gfg = graph.linearize(result,ord);
denseAb = gfg.denseJacobian

%% Get sparse matrix
IJS = gfg.sparse([12 7 4 10 1 14]');
Ab=sparse(IJS(1,:),IJS(2,:),IJS(3,:));
spy(Ab);

File: examples/PlanarSLAMExample_easy.cpp
/* ----------------------------------------------------------------------------

 * GTSAM Copyright 2010, Georgia Tech Research Corporation, 
 * Atlanta, Georgia 30332-0415
 * All Rights Reserved
 * Authors: Frank Dellaert, et al. (see THANKS for the full author list)

 * See LICENSE for the license information

 * -------------------------------------------------------------------------- */

/**
 * @file PlanarSLAMExample.cpp
 * @brief Simple robotics example using the pre-built planar SLAM domain
 * @author Alex Cunningham
 */

// pull in the planar SLAM domain with all typedefs and helper functions defined
#include <gtsam/slam/planarSLAM.h>

// we will use Symbol keys
#include <gtsam/nonlinear/Symbol.h>

using namespace std;
using namespace gtsam;

/**
 * Example of a simple 2D planar slam example with landmarls
 *  - The robot and landmarks are on a 2 meter grid
 *  - Robot poses are facing along the X axis (horizontal, to the right in 2D)
 *  - The robot moves 2 meters each step
 *  - We have full odometry between poses
 *  - We have bearing and range information for measurements
 *  - Landmarks are 2 meters away from the robot trajectory
 */
int main(int argc, char** argv) {

  // create the graph (defined in planarSlam.h, derived from NonlinearFactorGraph)
	planarSLAM::Graph graph;

	// Create some keys
	static Symbol i1('x',1), i2('x',2), i3('x',3);
	static Symbol j1('l',1), j2('l',2);

	// add a Gaussian prior on pose x_1
	Pose2 priorMean(0.0, 0.0, 0.0); // prior mean is at origin
	SharedDiagonal priorNoise(Vector_(3, 0.3, 0.3, 0.1)); // 30cm std on x,y, 0.1 rad on theta
	graph.addPrior(i1, priorMean, priorNoise); // add directly to graph

	// add two odometry factors
	Pose2 odometry(2.0, 0.0, 0.0); // create a measurement for both factors (the same in this case)
	SharedDiagonal odometryNoise(Vector_(3, 0.2, 0.2, 0.1)); // 20cm std on x,y, 0.1 rad on theta
	graph.addOdometry(i1, i2, odometry, odometryNoise);
	graph.addOdometry(i2, i3, odometry, odometryNoise);

	// create a noise model for the landmark measurements
	SharedDiagonal measurementNoise(Vector_(2, 0.1, 0.2)); // 0.1 rad std on bearing, 20cm on range

	// create the measurement values - indices are (pose id, landmark id)
	Rot2 bearing11 = Rot2::fromDegrees(45),
		   bearing21 = Rot2::fromDegrees(90),
		   bearing32 = Rot2::fromDegrees(90);
	double range11 = sqrt(4+4),
		     range21 = 2.0,
		     range32 = 2.0;

	// add bearing/range factors (created by "addBearingRange")
	graph.addBearingRange(i1, j1, bearing11, range11, measurementNoise);
	graph.addBearingRange(i2, j1, bearing21, range21, measurementNoise);
	graph.addBearingRange(i3, j2, bearing32, range32, measurementNoise);

	// print
	graph.print("Factor graph");

	// create (deliberatly inaccurate) initial estimate
	planarSLAM::Values initialEstimate;
	initialEstimate.insertPose(i1, Pose2(0.5, 0.0, 0.2));
	initialEstimate.insertPose(i2, Pose2(2.3, 0.1,-0.2));
	initialEstimate.insertPose(i3, Pose2(4.1, 0.1, 0.1));
	initialEstimate.insertPoint(j1, Point2(1.8, 2.1));
	initialEstimate.insertPoint(j2, Point2(4.1, 1.8));

	initialEstimate.print("Initial estimate:\n  ");

	// optimize using Levenberg-Marquardt optimization with an ordering from colamd
	planarSLAM::Values result = graph.optimize(initialEstimate);
	result.print("Final result:\n  ");

	return 0;
}

